experiments/pipeline/src/main/scala/org/apache/gearpump/experiments/pipeline/Processors.scala

```diff
@@ -0,0 +1,344 @@
+
+object PipelineConfig {
+  implicit def getConfig(pipelineConfig: PipelineConfig): Config = pipelineConfig.config
+}
+
+object DatumHandler {
+  val DEFAULT_INTERVAL = 2
+  val CPU = "CPU"
+  val CPU_INTERVAL = "pipeline.cpu.interval"
+  val MEM = "MEM"
+  val MEM_INTERVAL = "pipeline.memory.interval"
+
+  implicit def handle(data: String): Datum = {
+    read[Datum](data)
+  }
+  implicit def handle(datum: Datum): String = {
+    write[Datum](datum)
+  }
+}
+
+trait Average {
+  this: Task =>
+  var averageMem: Double = 0
+  var totalCount: Long = 0
+  var totalSum: Double = 0
+  var timeStamp: TimeStamp = 0
+
+  def timeInterval: Int
+
+  def average(datum: Datum)(implicit timeStamp: TimeStamp): Option[Message] = {
+    totalCount += 1
+    totalSum += datum.value
+    averageMem = totalSum/totalCount
+    interval(datum)
+  }
+
+  def elapsedInSec(implicit ts: Long): Long = ts - timeStamp
+
+  def interval(datum: Datum)(implicit ts: TimeStamp): Option[Message] = {
+    timeStamp match {
+      case 0 =>
+        timeStamp = ts
+        None
+      case _ =>
+        elapsedInSec match {
+          case delta if delta > timeInterval =>
+            timeStamp = ts
+            Some(Message(write[Datum](Datum(datum.dimension,datum.metric,averageMem)), timeStamp))
+          case _ =>
+            None
+        }
+    }
+  }
+}
+
+class CpuProcessor(taskContext: TaskContext, conf: UserConfig)
+  extends Task(taskContext, conf) with Average {
+  import taskContext.output
+
+  val pipelineConfig = taskConfig.getValue[PipelineConfig](PIPELINE)
+  override def timeInterval = pipelineConfig.map(config => {
+    config.getInt(CPU_INTERVAL)
+  }).getOrElse(DEFAULT_INTERVAL)
+
+  override def onStart(newStartTime: StartTime): Unit = {
+    LOG.info("starting timeInterval=$timeInterval")
+  }
+
+  override def onNext(msg: Message): Unit = {
+    Try({
+      implicit val timeStamp = msg.timestamp
+      val jsonData = msg.msg.asInstanceOf[String]
+      val envelope = read[Envelope](jsonData)
+      val body = read[Body](envelope.body)
+      val metrics = body.metrics
+      metrics.foreach(datum => {
+        datum.dimension match {
+          case CPU =>
+            average(datum).foreach(output)
+          case _ =>
+        }
+      })
+    }).failed.foreach(LOG.error("bad message", _))
+  }
+}
+
+class MemoryProcessor(taskContext: TaskContext, conf: UserConfig)
+  extends Task(taskContext, conf) with Average {
+  import taskContext.output
+
+  val pipelineConfig = taskConfig.getValue[PipelineConfig](PIPELINE)
+  override def timeInterval = pipelineConfig.map(config => {
+    config.getInt(MEM_INTERVAL)
+  }).getOrElse(DEFAULT_INTERVAL)
+
+  override def onStart(newStartTime: StartTime): Unit = {
+    LOG.info("starting timeInterval=$timeInterval")
+  }
+
+  override def onNext(msg: Message): Unit = {
+    Try({
+      implicit val timeStamp = msg.timestamp
+      val jsonData = msg.msg.asInstanceOf[String]
```

Comment:
Is it possible to avoid using upickle to pass data between processors?

You only need to create a implicit reference.conf under resource/, and define a default serializer for Datum

```
gearpump {
  serializers {
    "org.apache.gearpump...... Datum" = ""
  }
}
```
check https://github.com/intel-hadoop/gearpump/blob/master/experiments/storm/src/main/resources/reference.conf for an example